### DISCUSSION TITLE
[flat config] What is the recommended way to disable and override `globals`?

### DISCUSSION THREAD

**Discussion by mxxk** [eslint/eslint#17193] [Q&A] [answered]
In the `.eslintrc` system, it is possible to override entire groups of globals using `env`:

```js
// .eslintrc.js
module.exports = {
  root: true,
  extends: ["eslint:recommended"],
  parserOptions: {
    ecmaVersion: 2022,
  },
  // Assume all files access Node.js globals
  env: {
    node: true,
  },
  overrides: [
    {
      files: "file.js",
      // Except `file.js`, which only accesses browser globals
      env: {
        node: false,
        browser: true,
      },
    },
  ],
};
```

However, to express the same override using flat config, a "negated" globals entry must be provided:

```js
// eslint.config.js (flat config)
const eslintConfig = require("@eslint/js");
const globals = require("globals");

module.exports = [
  eslintConfig.configs.recommended,
  {
    languageOptions: {
      // Assume all files access Node.js globals
      globals: globals.node,
    },
  },
  {
    files: ["file.js"],
    languageOptions: {
      // Except `file.js`, which only accesses browser globals.
      globals: {
        // Explicitly disable Node.js globals
        ...Object.fromEntries(
          Object.keys(globals.node).map(key => [key, 'off']),
        ),
        ...globals.browser,
      },
    },
  },
];
```

**Is this recommended way to accomplish the above using flat config?** I suspect so, but wanted to get confirmation here.

### Additional notes

- There is symmetry between `node: false` and the "negated" `globals.node` expression—if either entry is omitted, then `file.js` will have access to _both_ Node.js and browser globals.
- If more than one set of globals needs to be disabled and overridden, then the `.eslintrc.js` equivalent would be less verbose than `eslint.config.js`, though the negation expression can be refactored into a helper function for better reuse.

**Comment by aladdin-add** [comment]
as I know there is no "recommended" way.  From my preference, I like to just enable the needed globals, not to disable. e.g.

```js
module.exports = [
  {languageOptions: {globals: commonGlobal}},
  {
    files: ['xxx.js'],  // the js run on browsers
    languageOptions: {globals: globals.browser},
  },
  {
    files: ['yyy.js'],  // the js run on node.js
    languageOptions: {globals: globals.node},
  },
];
```

**Reply by mxxk** [reply]
Thanks @aladdin-add! Though your suggestion to only have use additive `globals` works well for the minimal example posed in this question, I was specifically interested in using subtractive `globals` overrides. The real motivation behind this question is a far more complicated `.eslintrc` configuration which I'm trying to port to `eslint.config.js`. The configuration makes extensive use of `env` overrides. Refactoring it to only use additive `globals` would lead to many more config objects which otherwise share similar rules. Hence, I was curious specifically about subtractive overrides for `globals`.

**Reply by Rec0iL99** [reply, marked as answer]
Hi @mxxk, as @aladdin-add mentioned there is no "recommended" way, as the [ESLint blog](https://eslint.org/blog/2022/08/new-config-system-part-2/#goodbye-environments%2C-hello-globals) states ("...so we are handing this responsibility back to you."). I feel the community will come up with best practices on subtractive overrides once flat config starts to become mature :)

**Comment by johnloven** [comment]
This disables all globals except the ones you provide:

```js
function onlyGlobalsFrom(overrideWith) {
  return {
    ...Object.values(globals)
      .flatMap(Object.keys)
      .reduce((res, curr) => {
        res[curr] = "off";
        return res;
      }, {}),
    ...overrideWith,
  };
}
```

and can be used like

```js
languageOptions: {
  globals: onlyGlobalsFrom({ ...globals.builtin, ...globals.browser }),
}
```
